test/resilience/tests/envoygateway.go

```diff
@@ -0,0 +1,248 @@
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+//go:build resilience
+
+package tests
+
+import (
+	"context"
+	"github.com/envoyproxy/gateway/test/resilience/suite"
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/gateway-api/conformance/utils/http"
+	"sigs.k8s.io/gateway-api/conformance/utils/kubernetes"
+	"sigs.k8s.io/gateway-api/conformance/utils/tlog"
+	"testing"
+	"time"
+)
+
+const (
+	namespace    = "envoy-gateway-system"
+	envoygateway = "envoy-gateway"
+	targetString = "successfully acquired lease"
+	apiServerIP  = "10.96.0.1"
+	timeout      = 2 * time.Minute
+	policyName   = "egress-rules"
+	leaseName    = "5b9825d2.gateway.envoyproxy.io"
+	trashHold    = 2
+)
+
+func init() {
+	ResilienceTests = append(ResilienceTests, EGResilience)
+}
+
+var EGResilience = suite.ResilienceTest{
+	ShortName:   "EGResilience",
+	Description: "Envoygateway resilience test",
+	Test: func(t *testing.T, suite *suite.ResilienceTestSuite) {
+		ap := kubernetes.Applier{
+			ManifestFS:     suite.ManifestFS,
+			GatewayClass:   suite.GatewayClassName,
+			ControllerName: "gateway.envoyproxy.io/gatewayclass-controller",
+		}
+		ap.MustApplyWithCleanup(t, suite.Client, suite.TimeoutConfig, "testdata/base.yaml", true)
+
+		//this test will fail until https://github.com/envoyproxy/gateway/pull/4767/files is merged
+		t.Run("Secondary EnvoyGateway instances can serve an up to date xds", func(t *testing.T) {
+			ctx := context.Background()
+			t.Log("Scaling down the deployment to 0 replicas")
+			err := suite.Kube().ScaleDeploymentAndWait(ctx, envoygateway, namespace, 0, time.Minute, false)
+			require.NoError(t, err, "Failed to scale deployment replicas")
+
+			t.Log("Scaling up the deployment to 3 replicas")
+			err = suite.Kube().ScaleDeploymentAndWait(ctx, envoygateway, namespace, 3, time.Minute, false)
+			require.NoError(t, err, "Failed to scale deployment replicas")
+
+			t.Log("Waiting for leader election")
+			// Ensure leadership was taken
+			name, err := suite.Kube().GetElectedLeader(context.Background(), namespace, leaseName, metav1.Now(), time.Minute*2)
+			require.NoError(t, err, "unable to detect leader election")
+
+			t.Log("Marking the identified pod as leader")
+			suite.Kube().MarkAsLeader(namespace, name)
+
+			t.Log("Simulating API server connection failure for all pods")
+			err = suite.WithResCleanUp(ctx, t, func() (client.Object, error) {
+				return suite.Kube().ManageEgress(ctx, apiServerIP, namespace, policyName, true, map[string]string{})
+			})
+			require.NoError(t, err, "Failed to simulate API server connection failure")
+
+			// leader pod should go down, the standby remain
+			t.Log("Verifying deployment scales down to 2 replica")
+			err = suite.Kube().CheckDeploymentReplicas(ctx, envoygateway, namespace, 2, time.Minute)
```

Comment:
Question: Even though the leader pod is done due to losing connection to the API server, the replicas will still reach 3 because the deployment has been scaled to 3 replicas?

Reply:
You’re absolutely right to consider the intermediate state during the transition, but it’s important to clarify how Kubernetes handles pods that lose connectivity to the API server. When a leader pod loses connectivity to the API server, Kubernetes does not immediately terminate or stop the pod. Instead, the pod itself detects the loss of connectivity, initiates a graceful teardown process, and restarts to attempt to reconnect to the API server. 

The replica count for the deployment remains at 3 throughout the process.

Kubernetes does not schedule a new pod to replace the one that lost connectivity because the existing pod is not considered failed from Kubernetes’ perspective. It’s the responsibility of the application running inside the pod (e.g., the leader election logic) to handle reconnection attempts or restart itself.

Reply:
Thanks for the explanation!